src/lib/Microsoft.Health.Events/EventHubProcessor/EventProcessor.cs

```diff
@@ -0,0 +1,113 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Messaging.EventHubs;
+using Azure.Messaging.EventHubs.Consumer;
+using Azure.Messaging.EventHubs.Processor;
+using EnsureThat;
+using Microsoft.Health.Events.EventCheckpointing;
+using Microsoft.Health.Events.EventConsumers.Service;
+using Microsoft.Health.Events.Model;
+
+namespace Microsoft.Health.Events.EventHubProcessor
+{
+    public class EventProcessor
+    {
+        private IEventConsumerService _eventConsumerService;
+        private StorageCheckpointClient _checkpointClient;
+
+        public EventProcessor(IEventConsumerService eventConsumerService, StorageCheckpointClient checkpointClient)
+        {
+            _eventConsumerService = eventConsumerService;
+            _checkpointClient = checkpointClient;
+        }
+
+        public async Task RunAsync(EventProcessorClient processor, CancellationToken ct)
+        {
+            EnsureArg.IsNotNull(processor);
+
+            Task ProcessEventHandler(ProcessEventArgs eventArgs)
+            {
+                try
+                {
+                    if (eventArgs.HasEvent)
+                    {
+                        var evt = new Event(
+                            eventArgs.Partition.PartitionId,
+                            eventArgs.Data.Body,
+                            eventArgs.Data.Offset,
+                            eventArgs.Data.SequenceNumber,
+                            eventArgs.Data.EnqueuedTime.UtcDateTime,
+                            eventArgs.Data.SystemProperties);
+
+                        _eventConsumerService.ConsumeEvent(evt);
+                    }
+                    else
+                    {
+                        var evt = new MaximumWaitEvent(eventArgs.Partition.PartitionId, DateTime.UtcNow);
+                        _eventConsumerService.ConsumeEvent(evt);
+                    }
+                }
+                catch
+                {
+                    throw;
+                }
+
+                return Task.CompletedTask;
+            }
+
+            Task ProcessErrorHandler(ProcessErrorEventArgs eventArgs)
+            {
+                // todo: add an error processor
+                Console.WriteLine(eventArgs.Exception.Message);
+                return Task.CompletedTask;
+            }
+
+            async Task ProcessInitializingHandler(PartitionInitializingEventArgs initArgs)
```

Comment:
It is partly due to not passing eventArgs around.

But I also did not like eventArgs.UpdateCheckpointAsync() because we don't really want to checkpoint at the end of each flush.

So if we are not checkpointing at the end of each flush, we would need some way of keeping track of some previous eventArgs objects and then execute eventArgs.UpdateCheckpointAsync() when we feel it is appropriate to checkpoint. 

The other piece is using eventArgs.UpdateCheckpointAsync is that we don't have control over what keys/values get added to the checkpoint. Right now it adds the [offset and sequence number](https://github.com/Azure/azure-sdk-for-net/blob/82d737c996d086504d3ce78a604b0dfefd134273/sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/BlobCheckpointStore/BlobsCheckpointStore.cs#L369) but not an enqueuedDateTime which we also thought would be helpful for reporting the status/progress.

Reply:
Thank you for the explanation.

Personally I think keeping the last eventArgs you processed is fine.  It seems like a small cost to pay .  This is information we have to track in one way or the other.  The point about status/tracking process is a good one but I don't think we need to checkpoint .  This really should be a metric and we shouldn't need to rely on it being saved as a datetime for watermark purposes in order to get our metrics.  And we should be able to do with the event time to metric.

Would be interested in what others think here.

---
In reply to: [535501756](https://github.com/microsoft/iomt-fhir/pull/78#discussion_r535501756) [](ancestors = 535501756)

Reply:
Simplifying the codebase by using the built in checkpointing sounds like a resonable suggestion. If we find that the extra values being checkpointed (i..e enqueuedDateTime) don't provide value then we could explore a refactor.